src/node.rs

```diff
@@ -114,16 +135,250 @@ impl InMemoryNodeInner {
+                Default::default(),
+            )));
+        }
+
+        let gas_for_bytecodes_pubdata: u32 =
+            pubdata_for_factory_deps * (gas_per_pubdata_byte as u32);
+
+        // We are using binary search to find the minimal values of gas_limit under which the transaction succeeds
+        let mut lower_bound = 0;
+        let mut upper_bound = MAX_L2_TX_GAS_LIMIT as u32;
+
+        while lower_bound + ESTIMATE_GAS_ACCEPTABLE_OVERESTIMATION < upper_bound {
+            let mid = (lower_bound + upper_bound) / 2;
+            let try_gas_limit = gas_for_bytecodes_pubdata + mid;
+
+            let estimate_gas_result = self.estimate_gas_step(
+                l2_tx.clone(),
+                gas_per_pubdata_byte,
+                try_gas_limit,
+                l1_gas_price,
+                base_fee,
+            );
+            if estimate_gas_result.is_err() {
+                lower_bound = mid + 1;
+            } else {
+                upper_bound = mid;
+            }
+        }
+
+        let tx_body_gas_limit = cmp::min(
+            MAX_L2_TX_GAS_LIMIT as u32,
+            (upper_bound as f32 * ESTIMATE_GAS_SCALE_FACTOR) as u32,
+        );
+        let suggested_gas_limit = tx_body_gas_limit + gas_for_bytecodes_pubdata;
+
+        let estimate_gas_result = self.estimate_gas_step(
+            l2_tx.clone(),
+            gas_per_pubdata_byte,
+            suggested_gas_limit,
+            l1_gas_price,
+            base_fee,
+        );
+
+        match estimate_gas_result {
+            Err(_) => Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "Transaction is unexecutable".into(),
+                Default::default(),
+            ))),
+            Ok(_) => {
+                let overhead: u32 = derive_gas_estimation_overhead(
+                    suggested_gas_limit,
+                    gas_per_pubdata_byte as u32,
+                    tx.encoding_len(),
+                );
+
+                let full_gas_limit =
+                    match tx_body_gas_limit.overflowing_add(gas_for_bytecodes_pubdata + overhead) {
+                        (value, false) => value,
+                        (_, true) => {
+                            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                                "exceeds block gas limit".into(),
+                                Default::default(),
+                            )))
+                        }
+                    };
+
+                let fee = Fee {
+                    max_fee_per_gas: base_fee.into(),
+                    max_priority_fee_per_gas: 0u32.into(),
+                    gas_limit: full_gas_limit.into(),
+                    gas_per_pubdata_limit: gas_per_pubdata_byte.into(),
+                };
+                Ok(fee)
+            }
+        }
+    }
+
+    /// Runs fee estimation against a sandbox vm with the given gas_limit.
+    fn estimate_gas_step(
```

Comment:
I'd try to make this function static (by removing &self).

As we want to make sure that it doesn't do any 'permanent' modifications to the 'self's fields.

Reply:
also - I'd pass the StorageView as argument (rather than fork_storage).

Reply:
Updated to be a static function.

Attempted to pass `StorageView`, but since there's the loop and I do an additional check after the binary search is complete, I would need to implement `Clone` for `ForkStorage`. Because of this, I went the easier path of using `fork_storage` ref.

Out of curiosity, what's the benefit of passing `StorageView` vs `&ForkStorage`?

Reply:
StorageView 'writes' are not persisted in the underlying storage.
So you can 'safely' pass it to the function, and afterwards throw it away, knowing that the function didn't modify any database.